receiver/elasticapmreceiver/factory.go

```diff
@@ -0,0 +1,110 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package elasticapmreceiver // import "github.com/elastic/opentelemetry-collector-components/receiver/elasticapmreceiver"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/confighttp"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/receiver"
+
+	"github.com/elastic/opentelemetry-collector-components/receiver/elasticapmreceiver/internal/metadata"
+	"github.com/elastic/opentelemetry-collector-components/receiver/elasticapmreceiver/internal/sharedcomponent"
+)
+
+const (
+	defaultEndpoint = "localhost:8200"
+)
+
+// NewFactory creates a new factory for the elasticapm receiver.
+func NewFactory() receiver.Factory {
+	return receiver.NewFactory(
+		metadata.Type,
+		createDefaultConfig,
+		receiver.WithLogs(createLogsReceiver, metadata.LogsStability),
+		receiver.WithMetrics(createMetricsReceiver, metadata.MetricsStability),
+		receiver.WithTraces(createTracesReceiver, metadata.TracesStability),
+	)
+}
+
+// createDefaultConfig creates a default config with the endpoint set to port 8200.
+func createDefaultConfig() component.Config {
+	defaultServerConfig := confighttp.NewDefaultServerConfig()
+	defaultServerConfig.Endpoint = defaultEndpoint
+	return &Config{
+		ServerConfig: defaultServerConfig,
+	}
+}
+
+// createLogsReceiver creates a logs receiver with the given configuration.
+func createLogsReceiver(
+	_ context.Context,
+	set receiver.Settings,
+	cfg component.Config,
+	consumer consumer.Logs,
+) (receiver.Logs, error) {
+	oCfg := cfg.(*Config)
+	r, err := receivers.LoadOrStore(oCfg, func() (*elasticAPMReceiver, error) {
+		return newElasticAPMReceiver(oCfg, set)
+	})
+	if err != nil {
+		return nil, err
+	}
+	r.Unwrap().nextLogs = consumer
+	return r, nil
+}
+
+// createMetricsReceiver creates a metrics receiver with the given configuration.
+func createMetricsReceiver(
+	_ context.Context,
+	set receiver.Settings,
+	cfg component.Config,
+	consumer consumer.Metrics,
+) (receiver.Metrics, error) {
+	oCfg := cfg.(*Config)
+	r, err := receivers.LoadOrStore(oCfg, func() (*elasticAPMReceiver, error) {
+		return newElasticAPMReceiver(oCfg, set)
+	})
+	if err != nil {
+		return nil, err
+	}
+	r.Unwrap().nextMetrics = consumer
+	return r, nil
+}
+
+// createTracesReceiver creates a traces receiver with the given configuration.
+func createTracesReceiver(
+	_ context.Context,
+	set receiver.Settings,
+	cfg component.Config,
+	consumer consumer.Traces,
+) (receiver.Traces, error) {
+	oCfg := cfg.(*Config)
+	r, err := receivers.LoadOrStore(oCfg, func() (*elasticAPMReceiver, error) {
+		return newElasticAPMReceiver(oCfg, set)
+	})
+	if err != nil {
+		return nil, err
+	}
+	r.Unwrap().nextTraces = consumer
+	return r, nil
+}
+
+var receivers = sharedcomponent.NewMap[*Config, *elasticAPMReceiver]()
```

Comment:
It is so bothersome that some important packages in contrib are internal causing us to duplicate things 😞

Reply:
Indeed :(